File: backend/app/services/api_key_service.py
from typing import List, Optional
from datetime import datetime, timedelta, timezone
import logging

from app.repositories.api_key_repo import APIKeyRepository
from app.schemas.api_key_schema import (
    APIKeyCreate,
    APIKeyInDB,
    APIKeyResponse,
    APIKeyCreateResponse
)
from app.utilities.api_key_utils import generate_api_key
from app.config import settings

logger = logging.getLogger(__name__)


class APIKeyService:
    """Service for managing API keys"""
    
    def __init__(self, api_key_repo: APIKeyRepository):
        self.api_key_repo = api_key_repo
        
    async def create_api_key(
        self, 
        wallet_address: str, 
        api_key_data: APIKeyCreate
    ) -> APIKeyCreateResponse:
        """
        Create a new API key for a wallet address.
        
        Args:
            wallet_address: The wallet address creating the key
            api_key_data: The API key creation data
            
        Returns:
            The created API key with the actual key value
            
        Raises:
            ValueError: If wallet has too many active keys
        """
        # Check if API keys are enabled
        if not settings.api_key_auth_enabled:
            raise ValueError("API key authentication is not enabled")
            
        # Check existing key count
        active_keys = await self.api_key_repo.count_active_keys_for_wallet(wallet_address)
        if active_keys >= settings.api_key_max_per_wallet:
            raise ValueError(f"Maximum API keys limit reached ({settings.api_key_max_per_wallet})")
            
        # Generate API key
        api_key, key_hash = generate_api_key(wallet_address, settings.api_key_secret_key)
        
        # Set expiration if not provided
        expires_at = api_key_data.expires_at
        if not expires_at:
            expires_at = datetime.now(timezone.utc) + timedelta(days=settings.api_key_default_expiration_days)
            
        # Create API key record
        api_key_db = APIKeyInDB(
            key_hash=key_hash,
            wallet_address=wallet_address,
            name=api_key_data.name,
            permissions=api_key_data.permissions or settings.api_key_default_permissions,
            expires_at=expires_at,
            metadata=api_key_data.metadata,
            created_at=datetime.now(timezone.utc),
            is_active=True
        )
        
        # Save to database
        await self.api_key_repo.create_api_key(api_key_db)
        
        logger.info(f"API key created for wallet {wallet_address}")
        
        # Return response with the actual key
        return APIKeyCreateResponse(
            api_key=api_key,
            name=api_key_db.name,
            permissions=api_key_db.permissions,
            created_at=api_key_db.created_at,
            expires_at=api_key_db.expires_at,
            is_active=api_key_db.is_active,
            metadata=api_key_db.metadata,
            last_used_at=None
        )
    
    async def list_api_keys(self, wallet_address: str) -> List[APIKeyResponse]:
        """
        List all API keys for a wallet address.
        
        Args:
            wallet_address: The wallet address
            
        Returns:
            List of API keys (without the actual key values)
        """
        api_keys = await self.api_key_repo.get_api_keys_by_wallet(wallet_address)
        
        # Convert to response format (without key hash)
        return [
            APIKeyResponse(
                name=key.name,
                permissions=key.permissions,
                created_at=key.created_at,
                last_used_at=key.last_used_at,
                expires_at=key.expires_at,
                is_active=key.is_active,
                metadata=key.metadata
            )
            for key in api_keys
        ]
    
    async def revoke_api_key(self, wallet_address: str, key_name: str) -> bool:
        """
        Revoke an API key by name.
        
        Args:
            wallet_address: The wallet address
            key_name: The name of the key to revoke
            
        Returns:
            True if revoked, False if not found
        """
        # Find the key by name
        api_keys = await self.api_key_repo.get_api_keys_by_wallet(wallet_address)
        
        for key in api_keys:
            if key.name == key_name and key.is_active:
                # Deactivate the key
                success = await self.api_key_repo.deactivate_api_key(
                    key.key_hash, 
                    wallet_address
                )
                if success:
                    logger.info(f"API key '{key_name}' revoked for wallet {wallet_address}")
                return success
                
        return False
    
    async def update_permissions(
        self, 
        wallet_address: str, 
        key_name: str, 
        permissions: List[str]
    ) -> bool:
        """
        Update permissions for an API key.
        
        Args:
            wallet_address: The wallet address
            key_name: The name of the key to update
            permissions: New permissions list
            
        Returns:
            True if updated, False if not found
        """
        # Validate permissions
        valid_permissions = {"read", "write", "delete"}
        if not all(perm in valid_permissions for perm in permissions):
            raise ValueError("Invalid permissions")
            
        # Find the key by name
        api_keys = await self.api_key_repo.get_api_keys_by_wallet(wallet_address)
        
        for key in api_keys:
            if key.name == key_name and key.is_active:
                # Update permissions
                success = await self.api_key_repo.update_permissions(
                    key.key_hash,
                    wallet_address,
                    permissions
                )
                if success:
                    logger.info(f"Permissions updated for API key '{key_name}'")
                return success
                
        return False

File: frontend/src/services/apiKeyService.js
import apiClient from './apiClient';

const API_KEYS_BASE_URL = '/api-keys';

/**
 * API Key Service - handles all API key related operations
 */
const apiKeyService = {
  /**
   * Get API keys feature status
   * @returns {Promise<Object>} API keys status and configuration
   */
  getStatus: async () => {
    try {
      const response = await apiClient.get(`${API_KEYS_BASE_URL}/status`);
      return response.data;
    } catch (error) {
      console.error('Error getting API keys status:', error);
      throw error;
    }
  },

  /**
   * Create a new API key
   * @param {Object} apiKeyData - The API key creation data
   * @param {string} apiKeyData.name - User-friendly name for the API key
   * @param {string[]} apiKeyData.permissions - Array of permissions (read, write, delete)
   * @param {string} apiKeyData.expires_at - Optional expiration date
   * @param {Object} apiKeyData.metadata - Optional metadata
   * @returns {Promise<Object>} Created API key data including the key itself
   */
  createApiKey: async (apiKeyData) => {
    try {
      const response = await apiClient.post(`${API_KEYS_BASE_URL}/create`, apiKeyData);
      return response.data;
    } catch (error) {
      console.error('Error creating API key:', error);
      throw error;
    }
  },

  /**
   * List all API keys for the authenticated user
   * @returns {Promise<Array>} List of API keys (without the actual key values)
   */
  listApiKeys: async () => {
    try {
      const response = await apiClient.get(`${API_KEYS_BASE_URL}/list`);
      return response.data;
    } catch (error) {
      console.error('Error listing API keys:', error);
      throw error;
    }
  },

  /**
   * Revoke an API key
   * @param {string} keyName - The name of the API key to revoke
   * @returns {Promise<Object>} Success message
   */
  revokeApiKey: async (keyName) => {
    try {
      const response = await apiClient.delete(`${API_KEYS_BASE_URL}/${keyName}`);
      return response.data;
    } catch (error) {
      console.error('Error revoking API key:', error);
      throw error;
    }
  },

  /**
   * Update permissions for an API key
   * @param {string} keyName - The name of the API key
   * @param {string[]} permissions - New permissions array
   * @returns {Promise<Object>} Success message
   */
  updateApiKeyPermissions: async (keyName, permissions) => {
    try {
      const response = await apiClient.put(`${API_KEYS_BASE_URL}/${keyName}/permissions`, {
        permissions
      });
      return response.data;
    } catch (error) {
      console.error('Error updating API key permissions:', error);
      throw error;
    }
  },

  /**
   * Format permissions for display
   * @param {string[]} permissions - Array of permission strings
   * @returns {string} Formatted permissions string
   */
  formatPermissions: (permissions) => {
    if (!permissions || permissions.length === 0) return 'None';
    return permissions.map(p => p.charAt(0).toUpperCase() + p.slice(1)).join(', ');
  },

  /**
   * Check if a permission is valid
   * @param {string} permission - Permission to check
   * @returns {boolean} Whether the permission is valid
   */
  isValidPermission: (permission) => {
    const validPermissions = ['read', 'write', 'delete'];
    return validPermissions.includes(permission);
  },

  /**
   * Get available permissions
   * @returns {Array<{value: string, label: string}>} Available permissions
   */
  getAvailablePermissions: () => [
    { value: 'read', label: 'Read - View assets and data' },
    { value: 'write', label: 'Write - Create and update assets' },
    { value: 'delete', label: 'Delete - Remove assets' }
  ]
};

export default apiKeyService;